infrastructure/src/main/java/org/corfudb/infrastructure/management/ReconfigurationEventHandler.java

```diff
@@ -102,4 +116,32 @@ public static boolean handleHealing(@Nonnull CorfuRuntime runtime,
             return false;
         }
     }
+
+    /**
+     * Launches a workflow to restore redundancy and merge all segments.
+     *
+     * @param runtime Connected corfu runtime instance.
+     * @param layout  Latest known layout.
+     */
+    public static void handleMergeSegments(@Nonnull CorfuRuntime runtime, @Nonnull Layout layout) {
+
+        Duration workflowTimeout = getReasonableWorkflowTimeout(runtime);
+        log.info("triggerMergeSegments: Workflow to merge segments for layout {} timeout set to {} ms",
+                layout, workflowTimeout);
+
+        final Duration retryQueryTimeout = Duration.ofSeconds(1);
+
+        // Multiple invocations of this workflow can be avoided if the workflow are mapped to the same endpoint.
+        // Thus to deterministically choose an endpoint we always choose the last added log unit in the first stripe.
+        // Fetching the last segment. Note: This is the unbounded segment with ongoing writes.
+        LayoutSegment lastSegment = layout.getSegments().get(layout.getSegments().size() - 1);
+        // Fetching the first stripe for determinism.
+        List<String> firstStripeLogServers = lastSegment.getStripes().get(0).getLogServers();
+        runtime.getManagementView().mergeSegments(
+                // Last log unit in the stripe.
+                firstStripeLogServers.get(firstStripeLogServers.size() - 1),
```

Comment:
This should not be calculated here but should be in helper. Also if the layout changes between first and second try then are we just doomed for that run ? If so why can't we do better.